### PR TITLE
Baris/infer and evaluate

### PR DESCRIPTION
**Pull request by bcdurak**
## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [X] I have read the **CONTRIBUTING.md** document.
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

This is one of the significant updates which affects some of the most fundamental components. In summary, this update will allow our users to use an "AgnosticEvaluator" in their training pipelines, which is capable of evaluating regardless of the model origin.

List of changes:

- Implemented a new evaluator step called `AgnosticEvaluator` which takes the test results produced by a trainer step.
- Adjusted the base trainer step and the implemented trainer steps to include a `test_fn` which could be used as an interface to generate the aforementioned test results.
- Implemented an adapted version of the TFX Trainer component-executor pair to include a new `output_artifact` in the `fn_args`, so we can save the test results.
- Implemented an adapted version of the TFX Evaluator component which can operate without a given model and run `model_agnostic` evaluation
- Added an exemplary TensorBoard integration to the implemented PyTorch ff_trainer
- Added a new and more refined naming scheme on how to name the features which flow through the pipeline. It is implemented as a utility function and it is not necessary to use it while designing pipelines, it is built with the purpose of giving a good example.
- Adjusted the older module to work with the new changes and applied a large number of bug fixes

### CODE REVIEW

File: zenml/pipelines/training_pipeline.py [hunk 1]
```diff
@@ -204,23 +203,26 @@ def get_tfx_component_list(self, config: Dict[Text, Any]) -> List:
...
-                steps[keys.TrainingSteps.EVALUATOR])
-            eval_config = eval_step.build_eval_config()
-            evaluator = Evaluator(
-                examples=transform.outputs.transformed_examples,
-                model=trainer.outputs.model,
-                eval_config=eval_config,
-                module_file=custom_extractor_path,
-            ).with_id(GDPComponent.Evaluator.name)
+            evaluator_config = steps[keys.TrainingSteps.EVALUATOR]
+            # TODO: This check should ideally not happen here, integrate it
```
Comment (htahir1) [inline zenml/pipelines/training_pipeline.py:207]: Can you add [LOW], [MEDIUM], [HIGH] here?

File: zenml/components/evaluator/executor.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+_TELEMETRY_DESCRIPTORS = ['Evaluator']
+
+
+class Executor(base_executor.BaseExecutor):
+
+    def Do(self,
+           input_dict: Dict[Text, List[types.Artifact]],
+           output_dict: Dict[Text, List[types.Artifact]],
+           exec_properties: Dict[Text, Any]) -> None:
+
```
Comment (htahir1) [inline zenml/components/evaluator/executor.py:43]: maybe a docstring here to explain what this does?

File: zenml/steps/trainer/utils.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    path_utils.create_dir_if_not_exists(output_path)
+    output_path = os.path.join(output_path, 'test_results')
+    with tf.io.TFRecordWriter(output_path) as writer:
+        for example in to_serialized_examples(results):
+            writer.write(example)
+
+    return results
+
+
+def to_serialized_examples(instance) -> tf.train.Example:
```
Comment (htahir1) [inline zenml/steps/trainer/utils.py:21]: It looks like ive seen this sort of thing now a 100 times :-D can we consolidate into a utils? I think it is also in splitgen or datagen

File: zenml/components/bulk_inferrer/executor.py [hunk 4]
```diff
@@ -14,49 +14,33 @@
 """ZenML bulk inferrer executor."""
```
Comment (htahir1) [inline zenml/components/bulk_inferrer/executor.py:14]: Im not quite sure about the executor of the BulkInferrer anymore... Is it being used in the BatchInferencePipeline?

**Review by htahir1 — COMMENTED** [review]
Amazing job. I cant pretend to understand everything just yet. But commented on things I did understand and asked for some clarifications

File: zenml/pipelines/training_pipeline.py (hunk 1 above)
Comment (bcdurak) [reply zenml/pipelines/training_pipeline.py]: Added the MEDIUM tag

File: zenml/steps/trainer/utils.py (hunk 3 above)
Comment (bcdurak) [reply zenml/steps/trainer/utils.py:21]: Yes. Unfortunately, even though the resulting structure is a `tf.train.Example`, the structure of the input is different based on the operation, thus the computation is different. I am not 100% sure if we can come up with a single method, which would cover all of the instances.

**Pull request merged**